Chieftain abilities

Zorgovskiy opened the pull request:
What is missing:

Visuals for mounted state
Some sort of indication for the bottom of "Master the reins", possibly a character token on the summon

-------------------------------------------

Some things important to mention:
There are differences between FH style adaptation and the last official version on discord channel, the official version is taken.
Differences I noticed:
* In the official version Pipe Tomahawk initiative is 26, not 32
* In the official version War Paint bottom is 1xp, not 2
* In the official version Battle Rhinoceros is Unrecoverable
* In the official version Strapping Bullwhip has 1 xp at the last hex
* Sabretooth is Sabretooth, not Sabertooth!

--------------------------------------

As opposed to Granted/Control abilities by other characters that are Move X, Attack X
Chieftain grants Move +X, Attack +X
The difference is that basic summon stats are used, and the summons range and range type are applied as well ( even though there are no ranged summons at the moment )

Commit:
Initial chieftain files

Commit:
New traits and scenario events

Commit:
Chieftain assets and mount trait

Commit:
Merge branch 'develop' into chieftain-preparations

# Conflicts:
#	Game/Scripts/Models/Abilities/TargetedAbility.cs

Commit:
3 abilities and mounted check

Commit:
Change event type

Commit:
Improved mounted requirement

Commit:
More summons

Commit:
Remove builder and add more mounts and forcedmovementcheck

Commit:
Fixes to mount logic

Commit:
Improve owner logic

Commit:
Fix ability and make movement smooth

Commit:
Dismount dead summons

Commit:
More changes for chieftain

Commit:
Dismount dead again

Commit:
Mounted summon can open doors

Commit:
Fix target that was changed

Commit:
Merge branch 'chieftain-preparations' into chieftain-abilities

Commit:
Abilities+traits

Commit:
Fix regal beast and aoe patterns

Commit:
Support regal beast and add AOE trait

Commit:
Fixes after playtest 0

Commit:
Card fixes

Commit:
Fixes for Chieftain

Commit:
Fix AOEAttackTrait and DestroyObstacleTrait

Commit:
Fix War Elephant

Commit:
Add range and range type to granted attacks

Commit:
Merge branch 'develop' into chieftain-abilities

# Conflicts:
#	Game/Content/Classes/Chieftain/ChieftainModel.cs
#	Game/Scripts/Models/Abilities/AttackAbility.cs
#	Game/Scripts/Models/FigureTraits/AOEAttackTrait.cs
#	Game/Scripts/Models/FigureTraits/AllAttacksGainAdvantageTrait.cs
#	Game/Scripts/Models/FigureTraits/HealOnKillTrait.cs
#	Game/Scripts/Models/FigureTraits/TrapDisarmTrait.cs
#	Game/Scripts/Scenario/ScenarioEvents/ScenarioEvents.cs
#	Game/Scripts/UI/Popups/CreateCharacterPopup/CreateCharacterPopup.cs

Commit:
Revert unwanted changes

Commit:
Fixes due to trait changes

Commit:
Fixes due to trait changes

Commit:
Enable chieftain in campaign and adjust range

Commit:
Fixes and comments

Commit:
Add info to Master the reins

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/02_PiercingDarts.cs:
```diff
@@ -0,0 +1,74 @@
+						parameters => parameters.Performer == state.Performer,
+						async parameters =>
+						{
+							ScenarioCheckEvents.IsMountedCheck.Parameters isMountedCheckParameters =
```
You can create a helper method in AbilityCMD or in a Chieftain Specific class for this, as it's called quite a few times in different cards.

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/02_PiercingDarts.cs:
```diff
@@ -0,0 +1,74 @@
+				.WithRange(3)
+				.WithTargets(2)
+				.WithPierce(2)
+				.WithAfterTargetConfirmedSubscription(
```
I think this can be a during attack subscription rather than after target confirmed.

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/04_Resurrection.cs:
```diff
@@ -0,0 +1,59 @@
+					ScenarioEvents.DuringAttack.Subscription.ConsumeElement(Element.Earth,
+						applyFunction: async parameters =>
+						{
+							parameters.AbilityState.SingleTargetAdjustAttackValue(1);
```
I think this can be added to the full ability, rather than the single target.

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/09_SoulWhisperer.cs:
```diff
@@ -0,0 +1,72 @@
+							{
+								parameters.AbilityState.SingleTargetAdjustAttackValue(((Summon)parameters.Performer).Stats.Attack ?? 0);
+
+								int range = ((Summon)parameters.Performer).Stats.Range ?? 1;
```
Hmmm, perhaps we can think of a way to make this happen automatically. Maybe in the AttackAbility Builder, we can have a method WithGrantedPlusX(int x), which then handles everything needed. 🤔

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/11_HuntersMark.cs:
```diff
@@ -0,0 +1,120 @@
+			new AbilityCardAbility(OtherActiveAbility.Builder()
+				.WithOnActivate(async state =>
+				{
+					// TODO Place character token
```
*NITPICK ALERT* Could you place a `:` after the `TODO`? Otherwise I might miss it at some point when searching for `TODO:` 😅

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/09_SoulWhisperer.cs:
```diff
@@ -0,0 +1,72 @@
+							parameters => parameters.Performer == grantState.Target,
+							async parameters =>
+							{
+								parameters.AbilityState.SingleTargetAdjustAttackValue(((Summon)parameters.Performer).Stats.Attack ?? 0);
```
I think these adjustments can be ability wide, rather than single target.

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/15_TakeTheReins.cs:
```diff
@@ -0,0 +1,79 @@
+							parameters => parameters.Performer == grantState.Target,
+							async parameters =>
+							{
+								parameters.AbilityState.SingleTargetAdjustAttackValue(((Summon)parameters.Performer).Stats.Attack ?? 0);
```
I think these adjustments can be ability wide, rather than single target.

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/09_SoulWhisperer.cs:
```diff
@@ -0,0 +1,72 @@
+						.WithDistance(0)
+						.WithOnAbilityStarted(async moveState =>
+						{
+							moveState.AdjustMoveValue(((Summon)moveState.Performer).Stats.Move ?? 0);
```
Same as with the attack ability, perhaps we can think of a way to make this happen automatically. Maybe in the MoveAbility Builder, we can have a method WithGrantedPlusX(int x), which then handles everything needed. 🤔

Zorgovskiy replied:
> Same as with the attack ability, perhaps we can think of a way to make this happen automatically. Maybe in the MoveAbility Builder, we can have a method WithGrantedPlusX(int x), which then handles everything needed. 🤔

Pull request that extends dynamic values options allows doing that quite nicely, if it doesn't make it in I will think of another solution.

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/21_OneWithNature.cs:
```diff
@@ -0,0 +1,41 @@
+					ScenarioEvents.DuringAttack.Subscription.ConsumeElement(Element.Earth,
+						applyFunction: async applyParameters =>
+						{
+							applyParameters.AbilityState.SingleTargetAdjustAttackValue(2);
```
This can be for the entire ability, not just a single target.

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/22_VenomousMayhem.cs:
```diff
@@ -0,0 +1,44 @@
+			new AbilityCardAbility(SummonAbility.Builder()
+				.WithSummonStats(new SummonStats()
+				{
+					Health = 4,
```
This summon should have 5 health.

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/24_StrappingBullwhip.cs:
```diff
@@ -0,0 +1,83 @@
+							int distance = RangeHelper.Distance(parameters.Performer.Hex, parameters.AbilityState.Target.Hex);
+							parameters.AbilityState.SingleTargetAdjustAttackValue(distance);
+
+							if(distance == 3)
```
We can probably just leave this here for now, but there is a Frankenstein situation possible where a target gets added at this range, and it would give another XP. A way to fix this is to add an optional tag of sorts to the AOEHex class, which you can then compare with here to check whether or not to give XP. Since the Starslinger is also dependent on this kinda stuff (and I haven't looked at what Dareth has done for this), it's probably wise to implement this at some point.

HoogeboomBas replied:
Looking at the Starslinger cards with the updated design, it seems these XP for given hexes has been removed there as well. This is probably because Isaac himself doesn't like that in GH, so it's been taken out of all new designs for cards. I think we can therefore just remove this distance XP thing.

Zorgovskiy replied:
> Looking at the Starslinger cards with the updated design, it seems these XP for given hexes has been removed there as well. This is probably because Isaac himself doesn't like that in GH, so it's been taken out of all new designs for cards. I think we can therefore just remove this distance XP thing.

For starslinger it is quite a lot of cards, I think that can affect the XP generation quite a lot. For chieftain it is only 1 card, but I don't think it is a great solution just to remove that from everywhere.

HoogeboomBas replied:
Totally agree, but Starslinger does get XP on those cards, they XP is just not dependent on the yellow hexes anymore. From what I can see of those cards, it seems to be handled pretty well, and seems alright balance-wise.

HoogeboomBas commented on Game/Content/Classes/Chieftain/Cards/28_RegalBeast.cs:
```diff
@@ -0,0 +1,101 @@
+
+					foreach(AbilityCardSide abilitySide in abilitySidesToPerform)
+					{
+						await abilitySide.Perform(state.Performer);
```
I'm not a 100% sure on this, but can a user skip the summon ability here, after which the card remains in their hand?

Zorgovskiy replied:
> I'm not a 100% sure on this, but can a user skip the summon ability here, after which the card remains in their hand?

I don't know what is intended, couldn't find anything in the FAQ.
But if we are going with "Can skip anything non negative" then it is not correct right now. 
I will try to make it optional

HoogeboomBas replied:
Oh I was specifically wondering here what would happen if the player selects to skip here, because it seems they can skip the summon action on each of these cards, right? Since it states above that the selected cards are all returned to hand, and are then played, I wonder if they try to skip the ability, the cards go to the discard pile rather than getting set as an active persistent loss.

Zorgovskiy replied:
> Oh I was specifically wondering here what would happen if the player selects to skip here, because it seems they can skip the summon action on each of these cards, right? Since it states above that the selected cards are all returned to hand, and are then played, I wonder if they try to skip the ability, the cards go to the discard pile rather than getting set as an active persistent loss.

Oops you are right! It does go to the discard pile.
Should we just put the card to lost if it was skipped?

Zorgovskiy replied:
> Oh I was specifically wondering here what would happen if the player selects to skip here, because it seems they can skip the summon action on each of these cards, right? Since it states above that the selected cards are all returned to hand, and are then played, I wonder if they try to skip the ability, the cards go to the discard pile rather than getting set as an active persistent loss.

So technically the card says just play the cards, so if for some reason Chieftain gets a non-loss summon it should go to discard? But I didn't bother with it, just Lost all the recovered cards after they are played.

Commit:
Some review comments and a bugfix

Commit:
Updated spread

Commit:
Merge branch 'develop' into chieftain-abilities

Commit:
Review comments

Zorgovskiy commented:
I addressed everything except the +x to summon move attack/range stuff.
I can address it separately, either using the mechanism I introduced using dynamics values, or in other way if that is not the way to go.

Commit:
Change single target to ability

HoogeboomBas commented on Game/Content/Classes/Chieftain/Chieftain.cs:
```diff
@@ -0,0 +1,14 @@
+public partial class Chieftain : Character
```
Although this script currently only has static methods, since it inherits from Character, you can add it to the Chieftain.tscn scene. Just open the scene in the Godot editor and drag the script and drop it on the top Node, replacing the default Character script on there.

Commit:
Fix typo + add scene script

Commit:
Fix regal beast skipping summons

Zorgovskiy commented on Game/Content/Classes/Chieftain/Cards/28_RegalBeast.cs:
```diff
@@ -0,0 +1,118 @@
+					foreach(AbilityCardSide abilitySide in abilitySides)
+					{
+						await abilitySide.Perform(state.Performer);
+						await abilitySide.AbilityCard.SetCardState(abilitySide.AbilityCard.Unrecoverable ? CardState.UnrecoverablyLost : CardState.Lost);
```
User just recovered these cards, so they shouldn't be unrecoverable... but in any case

Commit:
Scene fix
